File: PitBoard(Android)/app/src/main/java/com/pitboard/app/data/AppSettingsRepository.kt
package com.pitboard.app.data

import android.content.Context
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.core.stringSetPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

private val Context.appSettingsDataStore by preferencesDataStore(name = "app_settings")

enum class AppTheme { LIGHT, DARK, SYSTEM }

class AppSettingsRepository(private val context: Context) {

    val notificationsEnabled: Flow<Boolean> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_NOTIFICATIONS_ENABLED] ?: true }

    /** Se marca la primera vez que la app pide el permiso del sistema (al abrirla por
     *  primera vez). Sirve para dos cosas: no volver a lanzar el dialogo en cada arranque
     *  y saber si un "denegado" ya es definitivo (ver NotificationPermission). */
    val notificationPermissionRequested: Flow<Boolean> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_NOTIFICATION_PERMISSION_REQUESTED] ?: false }

    val notificationMinutesBefore: Flow<Int> = context.appSettingsDataStore.data
        .map { prefs ->
            val stored = prefs[KEY_NOTIFICATION_MINUTES] ?: DEFAULT_MINUTES_BEFORE
            if (stored in VALID_MINUTES) stored else DEFAULT_MINUTES_BEFORE
        }

    val competitiveNotificationsEnabled: Flow<Boolean> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_COMPETITIVE_NOTIFICATIONS_ENABLED] ?: true }

    val practiceNotificationsEnabled: Flow<Boolean> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_PRACTICE_NOTIFICATIONS_ENABLED] ?: false }

    /** Clasificaciones (standings): a diferencia de los avisos, por defecto está
     *  DESACTIVADO — requiere internet (primera vez que la app sale a la red) y es un
     *  opt-in explícito, no algo que deba activarse solo. */
    val standingsEnabled: Flow<Boolean> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_STANDINGS_ENABLED] ?: false }

    val appTheme: Flow<AppTheme> = context.appSettingsDataStore.data
        .map { prefs ->
            prefs[KEY_APP_THEME]?.let { runCatching { AppTheme.valueOf(it) }.getOrNull() } ?: AppTheme.SYSTEM
        }

    /** Series sin avisos. Vacío = ninguna excluida (todas notifican). */
    val notificationDisabledSeries: Flow<Set<RaceSeries>> = context.appSettingsDataStore.data
        .map { prefs ->
            prefs[KEY_NOTIFICATION_DISABLED_SERIES]
                ?.asSequence()
                ?.mapNotNull { runCatching { RaceSeries.valueOf(it) }.getOrNull() }
                ?.toSet()
                ?: emptySet()
        }

    /** Series activas en el filtro rápido de Eventos. Vacío = todas (ver EventsScreen). */
    val eventScreenActiveSeries: Flow<Set<RaceSeries>> = context.appSettingsDataStore.data
        .map { prefs ->
            prefs[KEY_EVENT_SCREEN_SERIES]
                ?.asSequence()
                ?.mapNotNull { runCatching { RaceSeries.valueOf(it) }.getOrNull() }
                ?.toSet()
                ?: emptySet()
        }

    /** Tipos de sesión activos en el filtro rápido de Eventos (valores de SessionBadgeType:
     *  "C"/"Q"/"S"/"L"). Vacío = todos (mismo convenio que eventScreenActiveSeries). */
    val eventScreenActiveSessionTypes: Flow<Set<String>> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_EVENT_SCREEN_SESSION_TYPES] ?: emptySet() }

    /** 03/09/2026: se marca la primera vez que la app termina (con éxito o sin él) su
     *  sincronización de arranque de Eventos + Clasificaciones — ver MainActivity.
     *  Mientras esté en `false`, cada apertura de la app repite esa sincronización
     *  completa con pantalla de carga; en cuanto se marca, las siguientes aperturas ya
     *  no tocan Eventos (se queda con el ciclo diario en segundo plano) y solo
     *  refrescan Clasificaciones si el interruptor de Ajustes está activado. */
    val hasCompletedFirstSync: Flow<Boolean> = context.appSettingsDataStore.data
        .map { prefs -> prefs[KEY_HAS_COMPLETED_FIRST_SYNC] ?: false }

    suspend fun setNotificationsEnabled(enabled: Boolean) {
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_NOTIFICATIONS_ENABLED] = enabled }
    }

    suspend fun setNotificationPermissionRequested(requested: Boolean) {
        context.appSettingsDataStore.edit { prefs ->
            prefs[KEY_NOTIFICATION_PERMISSION_REQUESTED] = requested
        }
    }

    suspend fun setNotificationMinutesBefore(minutes: Int) {
        require(minutes in VALID_MINUTES) { "Antelación no soportada: $minutes" }
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_NOTIFICATION_MINUTES] = minutes }
    }

    suspend fun setCompetitiveNotificationsEnabled(enabled: Boolean) {
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_COMPETITIVE_NOTIFICATIONS_ENABLED] = enabled }
    }

    suspend fun setPracticeNotificationsEnabled(enabled: Boolean) {
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_PRACTICE_NOTIFICATIONS_ENABLED] = enabled }
    }

    suspend fun setStandingsEnabled(enabled: Boolean) {
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_STANDINGS_ENABLED] = enabled }
    }

    suspend fun setAppTheme(theme: AppTheme) {
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_APP_THEME] = theme.name }
    }

    suspend fun setNotificationDisabledSeries(series: Set<RaceSeries>) {
        context.appSettingsDataStore.edit { prefs ->
            if (series.isEmpty()) {
                prefs.remove(KEY_NOTIFICATION_DISABLED_SERIES)
            } else {
                prefs[KEY_NOTIFICATION_DISABLED_SERIES] = series.asSequence().map { it.name }.toSet()
            }
        }
    }

    suspend fun setEventScreenActiveSeries(series: Set<RaceSeries>) {
        context.appSettingsDataStore.edit { prefs ->
            prefs[KEY_EVENT_SCREEN_SERIES] = series.asSequence().map { it.name }.toSet()
        }
    }

    suspend fun setEventScreenActiveSessionTypes(sessionTypes: Set<String>) {
        context.appSettingsDataStore.edit { prefs ->
            prefs[KEY_EVENT_SCREEN_SESSION_TYPES] = sessionTypes
        }
    }

    suspend fun setHasCompletedFirstSync(completed: Boolean) {
        context.appSettingsDataStore.edit { prefs -> prefs[KEY_HAS_COMPLETED_FIRST_SYNC] = completed }
    }

    suspend fun isNotificationsEnabledNow(): Boolean = notificationsEnabled.first()

    suspend fun wasNotificationPermissionRequestedNow(): Boolean =
        notificationPermissionRequested.first()

    suspend fun isStandingsEnabledNow(): Boolean = standingsEnabled.first()

    suspend fun hasCompletedFirstSyncNow(): Boolean = hasCompletedFirstSync.first()

    suspend fun notificationDisabledSeriesNow(): Set<RaceSeries> =
        notificationDisabledSeries.first()

    companion object {
        const val DEFAULT_MINUTES_BEFORE = 60
        val VALID_MINUTES = setOf(15, 30, 60)

        private val KEY_NOTIFICATIONS_ENABLED = booleanPreferencesKey("notifications_enabled")
        private val KEY_NOTIFICATION_MINUTES = intPreferencesKey("notification_minutes_before")
        private val KEY_NOTIFICATION_PERMISSION_REQUESTED =
            booleanPreferencesKey("notification_permission_requested")
        private val KEY_COMPETITIVE_NOTIFICATIONS_ENABLED = booleanPreferencesKey("competitive_notifications_enabled")
        private val KEY_PRACTICE_NOTIFICATIONS_ENABLED = booleanPreferencesKey("practice_notifications_enabled")
        private val KEY_STANDINGS_ENABLED = booleanPreferencesKey("standings_enabled")
        private val KEY_APP_THEME = stringPreferencesKey("app_theme")
        private val KEY_NOTIFICATION_DISABLED_SERIES =
            stringSetPreferencesKey("notification_disabled_series")
        private val KEY_EVENT_SCREEN_SERIES = stringSetPreferencesKey("event_screen_series")
        private val KEY_EVENT_SCREEN_SESSION_TYPES = stringSetPreferencesKey("event_screen_session_types")
        private val KEY_HAS_COMPLETED_FIRST_SYNC = booleanPreferencesKey("has_completed_first_sync")
    }
}

File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Data/AppSettingsRepository.swift
import Foundation
import Observation

public enum AppTheme: String, Codable, Sendable, CaseIterable {
    case light = "LIGHT"
    case dark = "DARK"
    case system = "SYSTEM"
}

/// Preferencias del usuario — equivalente exacto de `AppSettingsRepository.kt`
/// (DataStore). Usa `UserDefaults(suiteName:)` sobre el mismo App Group que la base de
/// datos: a diferencia de Android (donde DataStore vive solo en el proceso de la app),
/// aquí el widget corre en un proceso separado y también necesita leer, como mínimo, el
/// tema y qué series están activas — compartir el suite evita duplicar estado.
///
/// Reactividad: en vez de exponer cada preferencia como un `Flow` (necesario en Android
/// para que Compose se recomponga), esta clase es `@Observable` — cualquier vista SwiftUI
/// que lea una de estas propiedades se actualiza sola cuando cambia, sin envoltorio
/// adicional. Las escrituras son funciones síncronas (UserDefaults ya es E/S local
/// rápida); se mantienen como funciones con nombre explícito (`setX`), igual que los
/// `suspend fun setX` de Kotlin, en vez de `didSet` implícito, para que quede clara la
/// intención en cada punto de llamada.
@Observable
public final class AppSettingsRepository: @unchecked Sendable {

    public static let defaultMinutesBefore = 60
    public static let validMinutes: Set<Int> = [15, 30, 60]

    private let defaults: UserDefaults

    public private(set) var notificationsEnabled: Bool
    public private(set) var notificationPermissionRequested: Bool
    public private(set) var notificationMinutesBefore: Int
    public private(set) var competitiveNotificationsEnabled: Bool
    public private(set) var practiceNotificationsEnabled: Bool
    /// A diferencia de los avisos, por defecto está DESACTIVADO — requiere internet y es
    /// opt-in explícito, igual que en Android.
    public private(set) var standingsEnabled: Bool
    public private(set) var appTheme: AppTheme
    /// Series sin avisos. Vacío = ninguna excluida (todas notifican).
    public private(set) var notificationDisabledSeries: Set<RaceSeries>
    /// Series activas en el filtro rápido de Eventos. Vacío = todas.
    public private(set) var eventScreenActiveSeries: Set<RaceSeries>
    /// Tipos de sesión activos en el filtro rápido de Eventos (rawValue de
    /// SessionBadgeType). Vacío = todos.
    public private(set) var eventScreenActiveSessionTypes: Set<String>
    /// Se marca cuando termina (con éxito o sin él) la primera sincronización de arranque
    /// de Eventos + Clasificaciones — mientras sea `false`, cada apertura repite esa
    /// sincronización completa con pantalla de carga (ver MainActivity en Android).
    public private(set) var hasCompletedFirstSync: Bool

    public init(suiteName: String = AppDatabase.appGroupId) {
        let defaults = UserDefaults(suiteName: suiteName) ?? .standard
        self.defaults = defaults

        self.notificationsEnabled = defaults.object(forKey: Keys.notificationsEnabled) as? Bool ?? true
        self.notificationPermissionRequested = defaults.bool(forKey: Keys.notificationPermissionRequested)
        let storedMinutes = defaults.object(forKey: Keys.notificationMinutes) as? Int ?? Self.defaultMinutesBefore
        self.notificationMinutesBefore = Self.validMinutes.contains(storedMinutes) ? storedMinutes : Self.defaultMinutesBefore
        self.competitiveNotificationsEnabled = defaults.object(forKey: Keys.competitiveNotificationsEnabled) as? Bool ?? true
        self.practiceNotificationsEnabled = defaults.object(forKey: Keys.practiceNotificationsEnabled) as? Bool ?? false
        self.standingsEnabled = defaults.object(forKey: Keys.standingsEnabled) as? Bool ?? false
        self.appTheme = (defaults.string(forKey: Keys.appTheme)).flatMap(AppTheme.init(rawValue:)) ?? .system
        self.notificationDisabledSeries = Set(
            (defaults.stringArray(forKey: Keys.notificationDisabledSeries) ?? []).compactMap(RaceSeries.init(rawValue:))
        )
        self.eventScreenActiveSeries = Set(
            (defaults.stringArray(forKey: Keys.eventScreenSeries) ?? []).compactMap(RaceSeries.init(rawValue:))
        )
        self.eventScreenActiveSessionTypes = Set(defaults.stringArray(forKey: Keys.eventScreenSessionTypes) ?? [])
        self.hasCompletedFirstSync = defaults.bool(forKey: Keys.hasCompletedFirstSync)
    }

    public func setNotificationsEnabled(_ enabled: Bool) {
        notificationsEnabled = enabled
        defaults.set(enabled, forKey: Keys.notificationsEnabled)
    }

    public func setNotificationPermissionRequested(_ requested: Bool) {
        notificationPermissionRequested = requested
        defaults.set(requested, forKey: Keys.notificationPermissionRequested)
    }

    public func setNotificationMinutesBefore(_ minutes: Int) {
        precondition(Self.validMinutes.contains(minutes), "Antelación no soportada: \(minutes)")
        notificationMinutesBefore = minutes
        defaults.set(minutes, forKey: Keys.notificationMinutes)
    }

    public func setCompetitiveNotificationsEnabled(_ enabled: Bool) {
        competitiveNotificationsEnabled = enabled
        defaults.set(enabled, forKey: Keys.competitiveNotificationsEnabled)
    }

    public func setPracticeNotificationsEnabled(_ enabled: Bool) {
        practiceNotificationsEnabled = enabled
        defaults.set(enabled, forKey: Keys.practiceNotificationsEnabled)
    }

    public func setStandingsEnabled(_ enabled: Bool) {
        standingsEnabled = enabled
        defaults.set(enabled, forKey: Keys.standingsEnabled)
    }

    public func setAppTheme(_ theme: AppTheme) {
        appTheme = theme
        defaults.set(theme.rawValue, forKey: Keys.appTheme)
    }

    public func setNotificationDisabledSeries(_ series: Set<RaceSeries>) {
        notificationDisabledSeries = series
        if series.isEmpty {
            defaults.removeObject(forKey: Keys.notificationDisabledSeries)
        } else {
            defaults.set(series.map(\.rawValue), forKey: Keys.notificationDisabledSeries)
        }
    }

    public func setEventScreenActiveSeries(_ series: Set<RaceSeries>) {
        eventScreenActiveSeries = series
        defaults.set(series.map(\.rawValue), forKey: Keys.eventScreenSeries)
    }

    public func setEventScreenActiveSessionTypes(_ sessionTypes: Set<String>) {
        eventScreenActiveSessionTypes = sessionTypes
        defaults.set(Array(sessionTypes), forKey: Keys.eventScreenSessionTypes)
    }

    public func setHasCompletedFirstSync(_ completed: Bool) {
        hasCompletedFirstSync = completed
        defaults.set(completed, forKey: Keys.hasCompletedFirstSync)
    }

    private enum Keys {
        static let notificationsEnabled = "notifications_enabled"
        static let notificationMinutes = "notification_minutes_before"
        static let notificationPermissionRequested = "notification_permission_requested"
        static let competitiveNotificationsEnabled = "competitive_notifications_enabled"
        static let practiceNotificationsEnabled = "practice_notifications_enabled"
        static let standingsEnabled = "standings_enabled"
        static let appTheme = "app_theme"
        static let notificationDisabledSeries = "notification_disabled_series"
        static let eventScreenSeries = "event_screen_series"
        static let eventScreenSessionTypes = "event_screen_session_types"
        static let hasCompletedFirstSync = "has_completed_first_sync"
    }
}
